src/bvh/bvh.rs

```diff
@@ -299,13 +366,13 @@ impl BoundingHierarchy for BVH {
     /// ```
     fn traverse<'a, T: Bounded>(&'a self, ray: &Ray, shapes: &'a [T]) -> Vec<&T> {
         let mut indices = Vec::new();
-        self.root.traverse_recursive(ray, &mut indices);
+        self.nodes[0].traverse_recursive(&self.nodes, ray, &mut indices);
```

Comment:
This looks bad. In terms of rust and in terms of design. `traverse_recursive` is canned on an element of the `self.nodes` array, while the array itself is passed to the function.
Modify the `BVHNode::traverse_recursive` function, make it static, and pass the index of the current node.